File: questions/1654. Minimum Jumps to Reach Home/min.py
class Solution:
    def minimumJumps(self, forbidden: List[int], a: int, b: int, x: int) -> int:
        max_v = max([x] + forbidden) + a + b
        arr = [0] + [float('inf')]*max_v
        for pos in forbidden: 
            arr[pos] = -1
        queue = deque()
        queue.append(0);
        while queue:
            cur = queue.popleft();
            if cur + a <= max_v and arr[cur + a] > arr[cur] + 1:
                arr[cur + a] = arr[cur] + 1
                queue.append(cur + a)
            if cur - b > 0 and arr[cur - b] > arr[cur] + 1:
                arr[cur - b] = arr[cur] + 1
                if cur - b + a <= max_v and arr[cur - b + a] > arr[cur] + 2:
                    arr[cur - b + a] = arr[cur] + 2
                    queue.append(cur - b + a)
        return arr[x] if arr[x] != float('inf') else -1
        

File: questions/1654. Minimum Jumps to Reach Home/min.cpp
class Solution {
public:
    int minimumJumps(vector<int>& forbidden, int a, int b, int x) {
        int max_v = max(*max_element(forbidden.begin(), forbidden.end()), x) + a + b;
        vector<int> arr(max_v + 1, INT_MAX);
        arr[0] = 0;
        for(int pos: forbidden){
            arr[pos] = -1;
        }
        queue<int> q;
        q.push(0);
        while(!q.empty()){
            int cur = q.front();
            q.pop();
            if(cur + a <= max_v && arr[cur + a] > arr[cur] + 1){
                arr[cur + a] = arr[cur] + 1;
                q.push(cur + a);
            }
            if(cur - b > 0 && arr[cur - b] > arr[cur] + 1){
                arr[cur - b] = arr[cur] + 1;
                if(cur - b + a <= max_v && arr[cur - b + a] > arr[cur] + 2){
                    arr[cur - b + a] = arr[cur] + 2;
                    q.push(cur - b + a);
                }
            }
        }
        return arr[x] == INT_MAX?-1:arr[x];
    }
};
